### PR TITLE
feat: replace Keycloak by FAM on FE

### LINKED ISSUE #480: Replace Keycloak by FAM for Authentication on FE
## Description

Replace Keycloak by FAM in the front end.

⚠️ Note that FAM needs to know beforehand all the URLs for redirecting, and that's not possible with PR's deployments. That means we'll not be able to log in into DEV, for now.

## Acceptance Criteria

- [ ] App running smoothly locally and on TEST
- [ ] Login with IDIR working
- [ ] Login with Business BCeID working
- [ ] All tests passing
- [ ] No Keycloak dependency left in the project

### PR DESCRIPTION
**Pull request by RMCampos**
# Description

Replace Keycloak by FAM.

Note that:
- Although this PR mess with `backend` this is only required to get it working until task #481 is finished.

Known issues:
- SignOut is taking too longer.
- When trying to go to a protected page (let's /seedlots) and the user **is** logged is, we can see the loading page. We on't want that. We should tackle this issue on a nother task.

Fixes #480 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Testes locally and with unit tests;

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged

## Further comments

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend](https://nr-spar-530-backend.apps.silver.devops.gov.bc.ca/)
[Frontend](https://nr-spar-530-frontend.apps.silver.devops.gov.bc.ca/)
[Oracle-API](https://nr-spar-530-oracle-api.apps.silver.devops.gov.bc.ca/)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend](https://nr-spar-530-backend.apps.silver.devops.gov.bc.ca/)
[Frontend](https://nr-spar-530-frontend.apps.silver.devops.gov.bc.ca/)
[Oracle-API](https://nr-spar-530-oracle-api.apps.silver.devops.gov.bc.ca/)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

### CODE REVIEW

**Comment by craigyu** [comment]
image

Cypress is failing because FAM returned error, is it due to the PR frontend url?

**Comment by DerekRoberts** [comment]
@craigyu I'm pretty sure there are issues with FAM and the PRs having different subdomains.  We can use different paths instead, if that solves the problem.  @franTarkenton solved a smilar issue with KeyCloak a while back by having domains added and removed in the workflows.  Any ideas?

**Comment by craigyu** [comment]
@DerekRoberts Not sure how using different paths will work but we can try, is it as secure?

**Comment by DerekRoberts** [comment]
> @DerekRoberts Not sure how using different paths will work but we can try, is it as secure?

@craigyu Security shouldn't change.  If FAM hates pr1.silva.gov.bc.ca, maybe it would like silva.gov.bc.ca/pr1 instead?  Just a suggestion, not necessarily a solution.

**Comment by RMCampos** [comment]
> image
> 
> Cypress is failing because FAM returned error, is it due to the PR frontend url?

Yup! Already asked to Catherine M. be adding this new PR URL. As soon that happen, we should be able to log in and out.

**Comment by RMCampos** [comment]
> > @DerekRoberts Not sure how using different paths will work but we can try, is it as secure?
> 
> @craigyu Security shouldn't change. If FAM hates pr1.silva.gov.bc.ca, maybe it would like silva.gov.bc.ca/pr1 instead? Just a suggestion, not necessarily a solution.

If I got it right, even a URL like that should be mapped before hand. No wildcards at all.

File: frontend/src/contexts/AuthProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    try {
+      await Auth.currentAuthenticatedUser();
+      userIsLoggedIn = true;
+      setSigned(userIsLoggedIn);
+    } catch (e) {}
+
+    if (userIsLoggedIn) {
+      const currentAuthToken: CognitoUserSession = await Auth.currentSession();
+      const famUser = parseToken(currentAuthToken);
+      localStorage.setItem(FAM_LOGIN_USER, JSON.stringify(famUser));
```
Comment (mgaseta) [inline frontend/src/contexts/AuthProvider.tsx:73]: Is it safe enough to leave this information on local storage? At a first glance, I think it's ok, but maybe we need to discuss a bit more.

File: frontend/src/index.tsx [hunk 2]
```diff
@@ -1,14 +1,15 @@
+window.global ||= window;
```
Comment (mgaseta) [inline frontend/src/index.tsx:1]: Had to do some research to understand this, but it makes sense! Just a suggestion, but maybe we can add it to another file and import it here. Like suggested [here](https://stackoverflow.com/questions/72114775/vite-global-is-not-defined/73208485#73208485).

**Review by mgaseta — CHANGES_REQUESTED** [review]
Really good job! :100: :rocket: 
Added only some questions and suggestions.

**Comment by RMCampos** [comment]
> > > @DerekRoberts Not sure how using different paths will work but we can try, is it as secure?
> > 
> > 
> > @craigyu Security shouldn't change. If FAM hates pr1.silva.gov.bc.ca, maybe it would like silva.gov.bc.ca/pr1 instead? Just a suggestion, not necessarily a solution.
> 
> If I got it right, even a URL like that should be mapped before hand. No wildcards at all.

@craigyu @DerekRoberts FAM was updated with PR URLs, but we need to update Caddy security config. Maybe Derek can help me on this one?
image

**Comment by mgaseta** [comment]
Also found this warning on the console, just thought it was worth mentioning.
Screenshot from 2023-10-03 10-13-13

File: backend/src/main/java/ca/bc/gov/backendstartapi/config/SecurityConfig.java [hunk 3]
```diff
@@ -39,6 +29,7 @@ public class SecurityConfig {
    */
   @Bean
   public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+    /*
```
[the remark this answers is not in the document]
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/config/SecurityConfig.java:32]: Yeah, probably habit. Removed here: https://github.com/bcgov/nr-spar/commit/78d2dbe4e4038e29f260974fb6fc0379834ea96a

File: backend/src/main/java/ca/bc/gov/backendstartapi/provider/OracleApiProvider.java [hunk 4]
```diff
@@ -137,10 +137,12 @@ public List<SameSpeciesTreeDto> findParentTreesByVegCode(
               requestEntity,
               new ParameterizedTypeReference<List<SameSpeciesTreeDto>>() {},
               createParamsMap("vegCode", vegCode));
+      List<SameSpeciesTreeDto> list = parentTreesResult.getBody();
+      int size = list == null ? 0 : list.size();
```
[the remark this answers is not in the document]
Comment (DerekRoberts) [reply backend/src/main/java/ca/bc/gov/backendstartapi/provider/OracleApiProvider.java:141]: I hate ternary!  It's always so hard to read and remember.

File: frontend/src/contexts/AuthProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+ */
+const parseToken = (authToken: CognitoUserSession): FamUser => {
+  const decodedIdToken = authToken.getIdToken().decodePayload();
+  const decodedAccessToken = authToken.getAccessToken().decodePayload();
+
+  // Extract the first name and last name from the displayName and remove unwanted part
+  const displayName = decodedIdToken['custom:idp_display_name'] as string;
+  const [lastName, firstName] = displayName.split(', ');
+  const sanitizedFirstName = firstName.split(' ')[0].trim(); // Remove unwanted part (WLRS:EX)
+  const sanitizedLastName = lastName.includes(' ') ? lastName.split(' ')[1].trim() : lastName;
```
[the remark this answers is not in the document]
Comment (RMCampos) [reply frontend/src/contexts/AuthProvider.tsx:36]: I don't think so. The last name will be a single word or in some cases, like my name "de Campos".

File: frontend/src/contexts/AuthProvider.tsx (hunk 1 above)
Comment (RMCampos) [reply frontend/src/contexts/AuthProvider.tsx:73]: Yeah, we can have a discussion on the best approach for that. We could use the backend and store a cookie for example.

File: frontend/src/index.tsx (hunk 2 above)
Comment (RMCampos) [reply frontend/src/index.tsx]: Good catch. Done here https://github.com/bcgov/nr-spar/commit/bcf86766c8b64f1010d64de8227042704f2c7b30

**Comment by craigyu** [comment]
image

I'm seeing this after logging out, probably another FAM issue.

**Comment by DerekRoberts** [comment]
> image
> 
> I'm seeing this after logging out, probably another FAM issue.

Could also be the logout redirect URL.  We can investigate tomorrow!  @craigyu 

@jazzgrewal did this, in case it's transferrable and relevant.
```
const redirectSignOut =
  env.VITE_REDIRECT_SIGN_OUT && env.VITE_REDIRECT_SIGN_OUT.trim() !== ""
    ? env.VITE_REDIRECT_SIGN_OUT
    : `${logoutDomain}/clp-cgi/logoff.cgi?retnow=1&returl=${retUrl}?redirect_uri=${redirectUri}/`;
```

https://github.com/bcgov/nr-results-exam/blob/789c098daeabacd9298958160d820413d72e85f1/frontend/src/aws-exports.js#L8

**Review by craigyu — APPROVED** [review]
Hello FAM!

**Comment by craigyu** [comment]
we can arguably remove `ProtectedRoute` altogether, but i don't mind the extra layer

**Comment by DerekRoberts** [comment]
Looks like we're getting close finally!  Just some SonarCloud coverage errors.  We have a parallel issue with Cypress coverage, so if that's the cause it's okay to skip or wait on that one.  @craigyu

**Comment by craigyu** [comment]
Closed to merge things into the staging branch

**Pull request closed without merging**